add routes to services view

pweil- opened the pull request:
@fabianofranz - this is step 1 of adding routes to the console.  It only adds them to the services page.  Could you give me a review and see if this is ok?  Still todo would be adding them to the box/diagram view.

pweil- commented:
I woke up at 3 am today and I realize that (I think) I need to maintain the list of routes separately from the sorted list in order to take advantage of the event type logic in `.by`.  I'll update that today.

fabianofranz commented:
@pweil- Reviewing
@jwforres @liggitt could you also take a look?

pweil- commented:
never mind my last comment, after a test it is working as I (didn't) expect.

fabianofranz commented:
LGTM

jwforres commented:
@pweil- can you include screenshots

jwforres commented on assets/app/scripts/controllers/services.js:
```diff
@@ -24,7 +25,24 @@ angular.module('openshiftConsole')
+    }));
+
+    watches.push(DataService.watch("routes", $scope, function(routes){
+        $scope.routes = sortRoutes(routes._data);
```
don't access the _data object directly

jwforres replied:
If sortRoutes is returning a map of routes grouped by service, then rename this to $scope.routesByService to be consistent with naming schemes in our other controllers

liggitt replied:
Actually, if `serviceName` will always be present on a route (at least as an empty string), you should be able to just do `$scope.routesByService = routes.by("serviceName");` and get rid of the `sortRoutes` function. (@jwforres, check me on that)

pweil- replied:
The `.by` function does not support multiple entries like we will have for routes (1 service name, many routes).  Will make the rest of the updates asap.  Thanks for the reviews!

jwforres commented on assets/app/scripts/controllers/services.js:
```diff
@@ -24,7 +25,24 @@ angular.module('openshiftConsole')
+        console.log("routes (subscribe)", $scope.routes);
+    }));
+
+    var sortRoutes = function(routes) {
```
also rename this to routesByService for consistency with other controllers

jwforres commented on assets/app/scripts/controllers/services.js:
```diff
@@ -24,7 +25,24 @@ angular.module('openshiftConsole')
+    var sortRoutes = function(routes) {
+        var routeMap = {};
+        angular.forEach(routes, function(route){
+            if (!routeMap[route.serviceName]){
```
we like to keep these reduced to one-liners:
<code>routeMap[route.serviceName] = routeMap[route.serviceName] || [];</code>

liggitt replied:
also, make it a double map for consistency

liggitt commented on assets/app/scripts/controllers/services.js:
```diff
@@ -24,7 +25,24 @@ angular.module('openshiftConsole')
+                routeMap[route.serviceName] = []
+            }
+
+            routeMap[route.serviceName].push(route)
```
make this a map, such that `routesByServe[route.serviceName][route.metadata.name] = route`

liggitt commented on assets/app/views/services.html:
```diff
@@ -16,7 +16,12 @@
         <div>Protocol: {{service.spec.protocol}}</div>
         <div>Selectors:<span ng-repeat="(selectorLabel, selectorValue) in service.spec.selector"> {{selectorLabel}}={{selectorValue}}</span></div>
-      </div>      
+        <div>Routes: <span ng-repeat="route in routes[service.metadata.name]"> {{route.host}}</span></div>
```
`{{route.host}}<span ng-show="!$last">, </span>`

liggitt replied:
`ng-repeat="(routeId, route) in routes[serviceId]"` (and change the `ng-repeat` above to `(serviceId, service) in services`)

liggitt commented on assets/app/views/services.html:
```diff
@@ -16,7 +16,12 @@
+      </div>
+
+      <div ng-if="routes['']">
+          <div>Routes With No Service: <span ng-repeat="route in routes['']"> {{route.host}}</span></div>
```
`{{route.host}}<span ng-show="!$last">, </span>`

liggitt replied:
`Routes With No Service` feels awkward to me. Try
```
<h3>No Service</h3>
<div>Routes: ...</div>
```

liggitt replied:
`ng-repeat="(routeId, route) in routes['']"`

liggitt commented on assets/app/views/services.html:
```diff
@@ -16,7 +16,12 @@
         <div>Container port: {{service.spec.containerPort}}</div>
         <div>Protocol: {{service.spec.protocol}}</div>
         <div>Selectors:<span ng-repeat="(selectorLabel, selectorValue) in service.spec.selector"> {{selectorLabel}}={{selectorValue}}</span></div>
```
@jwforres do we want comma-separation here as well? (e.g. `<span ng-show="!$last">, </span>`)

liggitt commented on assets/app/views/services.html:
```diff
@@ -16,7 +16,12 @@
+        <div>Routes: <span ng-repeat="route in routes[service.metadata.name]"> {{route.host}}</span></div>
+      </div>
+
+      <div ng-if="routes['']">
```
`ng-if=(routes[''] | hashSize)`

pweil- commented:
Updates based on feedback.  I should note, that right now you cannot have a route with no service so the second section will not be displayed (if anyone tries to test that part).  We will allow it in the future though so I think it's good to add it in anyway.  

@jwforres how do you feel about adding the comma separation on the selectors portion as suggested by @liggitt?  

![routes_on_services](https://cloud.githubusercontent.com/assets/5230419/6443053/558685bc-c0c2-11e4-8123-6d6a498fe477.png)

jwforres commented:
@pweil- yep make them comma separated, and when you don't have any selectors or dont have any routes for a service, can you add `<em>none</em>`

pweil- commented:
@jwforres 10-4, you got it!

liggitt commented:
@pweil- one other thought, do we know enough about the routes to know if they are webby (http/https) so we can linkify them?

pweil- commented:
Commafied and added /None/, PTAL

@liggitt - we're not guaranteed that the route is serving web traffic.  We really only know if it is an encrypted or unencrypted route.  I feel like most routes would be webby though so I don't see an issue with linking everything and letting someone yell if it's an issue.  @jwforres ?  

New screen shot for the use cases.

![routes_on_services](https://cloud.githubusercontent.com/assets/5230419/6445509/61c82474-c0d2-11e4-9a75-2f1f41bcc97d.png)

liggitt commented:
I'd like to make things links when it makes sense. One other question, should we include the route path when specified? Seems like we should.

liggitt commented:
I guess the http/https nature is planned to be part of the service definition, so we could only link routes in services we detect are webby (and make the "detection" function just return true for now)

jwforres commented:
sure, lets make the assumption for now and make them links and we'll revisit after services are re-done

jwforres commented on assets/app/views/services.html:
```diff
@@ -15,8 +15,14 @@
         <div>Protocol: {{service.spec.protocol}}</div>
-        <div>Selectors:<span ng-repeat="(selectorLabel, selectorValue) in service.spec.selector"> {{selectorLabel}}={{selectorValue}}</span></div>
-      </div>      
+        <div>Selectors: <span ng-if="!service.spec.selector || hashSize == 0"><em>None</em></span><span ng-repeat="(selectorLabel, selectorValue) in service.spec.selector"> {{selectorLabel}}={{selectorValue}}<span ng-show="!$last">, </span></span></div>
```
The upper-case N in none stands out a little too much, let's lower case it

pweil- commented:
Updated to lower case, added an isWebRoute return true function and a helper to craft the url.  

@liggitt I realized based on your filter comment that I wasn't using hashSize correctly, I updated those lines as well,  PTAL.  Keep in mind that if a service doesn't have routes it won't have an entry in the map so I couldn't always call hashSize (we'll get `.keys called on a non-object` errors).  So the logic should read, if you're in the map and you have a length > 0 show 'none' else show the routes.  Please let me know if I did it wrong.

pweil- commented:
New screen shot with links (the secure route is testing the https/http logic)

![routes_on_services](https://cloud.githubusercontent.com/assets/5230419/6447906/3541a1e8-c0e4-11e4-813d-ca3f5018d7f1.png)

liggitt commented on assets/app/views/services.html:
```diff
@@ -15,8 +15,16 @@
-        <div>Selectors:<span ng-repeat="(selectorLabel, selectorValue) in service.spec.selector"> {{selectorLabel}}={{selectorValue}}</span></div>
-      </div>      
+        <div>Selectors: <span ng-if="!service.spec.selector"><em>none</em></span><span ng-repeat="(selectorLabel, selectorValue) in service.spec.selector"> {{selectorLabel}}={{selectorValue}}<span ng-show="!$last">, </span></span></div>
+        <div>Routes: <span ng-if="(!routesByService[service.metadata.name]) || (routesByService[service.metadata.name] | hashSize) == 0"><em>none</em></span><span ng-if="(routesByService[service.metadata.name] | hashSize) > 0" ng-repeat="(routeName, route) in routesByService[service.metadata.name]">
```
if you pass hashSize an empty value, it can return 0 (so add `if (!hash) return 0` to hashSize), and that'll clean this up

liggitt replied:
Also, you don't need the `if > 0` test... repeating over an empty value will just never call the repeat

liggitt commented on assets/app/views/services.html:
```diff
@@ -15,8 +15,16 @@
-      </div>      
+        <div>Selectors: <span ng-if="!service.spec.selector"><em>none</em></span><span ng-repeat="(selectorLabel, selectorValue) in service.spec.selector"> {{selectorLabel}}={{selectorValue}}<span ng-show="!$last">, </span></span></div>
+        <div>Routes: <span ng-if="(!routesByService[service.metadata.name]) || (routesByService[service.metadata.name] | hashSize) == 0"><em>none</em></span><span ng-if="(routesByService[service.metadata.name] | hashSize) > 0" ng-repeat="(routeName, route) in routesByService[service.metadata.name]">
+            <a ng-if="route.url" href="{{route.url}}">{{route.host}}</a><span ng-if="!route.url">{{route.host}}</span><span ng-show="!$last">, </span></span>
```
line up the cases:
```
<a ng-if="(route | routeWebURL)" ...
<a ng-if="!(route | routeWebURL)" ...
```

liggitt commented on assets/app/scripts/controllers/services.js:
```diff
@@ -24,7 +25,35 @@ angular.module('openshiftConsole')
+        var routeMap = {};
+        angular.forEach(routes, function(route){
+            routeMap[route.serviceName] = routeMap[route.serviceName] || {};
+            route.url = getRouteUrl(route);
```
we try really hard not to mutate the objects we get back from the API, so that anyone who knows the API can work with the objects in the UI automatically. Maybe make the routeURL function a filter so it can be used in the view (and name it more descriptively, like `routeWebURL`)?

liggitt replied:
See scripts/filters/resources.js for examples

pweil- commented:
Updated to line up statements and general readability.  Update hashSize filter to return 0 if hash is not passed

liggitt commented on assets/app/scripts/controllers/services.js:
```diff
@@ -24,7 +25,35 @@ angular.module('openshiftConsole')
+
+    var routesByService = function(routes) {
+        var routeMap = {};
+        angular.forEach(routes, function(route){
```
you can take a second param here which is the `routeId` instead of looking up `route.metadata.name` below

pweil- commented:
@liggitt - added 2 filters, one is the placeholder function to determine if it is a web route or not (always returning true), the other is the url filter.  Updated the html to use new filters, removed the old code.

liggitt commented on assets/app/scripts/controllers/services.js:
```diff
@@ -24,7 +25,21 @@ angular.module('openshiftConsole')
+        var routeMap = {};
+        angular.forEach(routes, function(route, routeName){
+            routeMap[route.serviceName] = routeMap[route.serviceName] || {};
+            routeMap[route.serviceName][routeName] = route
```
needs trailing ';'

pweil- replied:
damn golang, it has left me with sloppy semi colons after all my mental retraining to not put them on every line.  Fix incoming

liggitt commented:
LGTM

liggitt commented:
@jwforres any last comments before merge?

jwforres commented:
[merge]
